actpool/queueworker.go

```diff
@@ -0,0 +1,286 @@
+		}
+	}()
+	return nil
+}
+
+func (worker *queueWorker) Stop() error {
+	close(worker.queue)
+	return nil
+}
+
+func (worker *queueWorker) Handle(job workerJob) error {
+	ctx := job.ctx
+	// ctx is canceled or timeout
+	if ctx.Err() != nil {
+		return ctx.Err()
+	}
+
+	var (
+		span            = tracer.SpanFromContext(ctx)
+		act             = job.act
+		sender          = act.SenderAddress().String()
+		actHash, _      = act.Hash()
+		intrinsicGas, _ = act.IntrinsicGas()
+	)
+	defer span.End()
+
+	nonce, balance, err := worker.getConfirmedState(ctx, act.SenderAddress())
+	if err != nil {
+		return err
+	}
+
+	if err := worker.checkSelpWithState(&act, nonce, balance); err != nil {
+		return err
+	}
+
+	if err := worker.putAction(sender, act, nonce, balance); err != nil {
+		return err
+	}
+
+	worker.ap.allActions.Set(actHash, act)
+
+	if desAddress, ok := act.Destination(); ok && !strings.EqualFold(sender, desAddress) {
+		worker.addDestinationMap(act)
+	}
+
+	atomic.AddUint64(&worker.ap.gasInPool, intrinsicGas)
+
+	worker.removeEmptyAccounts()
+
+	return nil
+}
+
+func (worker *queueWorker) getConfirmedState(ctx context.Context, sender address.Address) (uint64, *big.Int, error) {
+	worker.mu.RLock()
+	queue := worker.accountActs[sender.String()]
+	worker.mu.RUnlock()
+	// account state isn't cached in the actpool
+	if queue == nil {
+		confirmedState, err := accountutil.AccountState(ctx, worker.ap.sf, sender)
+		if err != nil {
+			return 0, nil, err
+		}
+		return confirmedState.PendingNonce(), confirmedState.Balance, nil
+	}
+	return queue.AccountNonce(), queue.AccountBalance(), nil
+}
+
+func (worker *queueWorker) checkSelpWithState(act *action.SealedEnvelope, pendingNonce uint64, balance *big.Int) error {
+	if act.Nonce() < pendingNonce {
+		_actpoolMtc.WithLabelValues("nonceTooSmall").Inc()
+		return action.ErrNonceTooLow
+	}
+
+	// Nonce exceeds current range
+	if act.Nonce()-pendingNonce >= worker.ap.cfg.MaxNumActsPerAcct {
+		hash, _ := act.Hash()
+		log.L().Debug("Rejecting action because nonce is too large.",
+			log.Hex("hash", hash[:]),
+			zap.Uint64("startNonce", pendingNonce),
+			zap.Uint64("actNonce", act.Nonce()))
+		_actpoolMtc.WithLabelValues("nonceTooLarge").Inc()
+		return action.ErrNonceTooHigh
+	}
+
+	if cost, _ := act.Cost(); balance.Cmp(cost) < 0 {
+		_actpoolMtc.WithLabelValues("insufficientBalance").Inc()
+		sender := act.SenderAddress().String()
+		actHash, _ := act.Hash()
+		log.L().Info("insufficient balance for action",
+			zap.String("actionHash", hex.EncodeToString(actHash[:])),
+			zap.String("cost", cost.String()),
+			zap.String("balance", balance.String()),
+			zap.String("sender", sender),
+		)
+		return action.ErrInsufficientFunds
+	}
+	return nil
+}
+
+func (worker *queueWorker) putAction(sender string, act action.SealedEnvelope, pendingNonce uint64, confirmedBalance *big.Int) error {
+	worker.mu.RLock()
+	queue := worker.accountActs[sender]
+	worker.mu.RUnlock()
+
+	if queue == nil {
```

Comment:
is it possible that two actions from the same user are put into queue at the same time when the queue is nil, as a result, one of the action is actually abandoned?

Reply:
Two acts in the queue are handled by the same thread sequentially at https://github.com/iotexproject/iotex-core/pull/3377/files/6b0a5f3041a99eef69eed90d32726bf22f9e0d97#diff-b09c50140126cf69760c5606ef46e97741ab1eadf95ca82d8c021c35c5efede2R60 ?